include/odp/api/spec/schedule.h

```diff
@@ -45,6 +45,94 @@ extern "C" {
  * Maximum schedule group name length in chars including null char
  */
 
+/**
+ * Schedule configuration
+ */
+typedef struct odp_schedule_config_t {
+	/** Number of flows per queue to be supported
+	 * Scheduler enables flow aware mode when flow count is configured
+	 * greater than 1 (up to capa.max_flows). 
+	 *
+	 * Flows are lightweight entities and events can be assigned to
+	 * specific flows by the application using odp_event_flow_id_set()
+	 * before enqueuing the event into the scheduler. This value is
+	 * ignored unless scheduler supports flow aware mode.
+	 *
+	 * Depending on the implementation this number might be rounded up to
+	 * nearest supported value (e.g power of 2). odp_schedule_config() will
+	 * update passed config with the value actually used. This number
+	 * should be less than maximum flow supported by the implementation.
+	 *
+	 * @see odp_schedule_capability_t
+	 */
+	uint32_t num_flows;
+
+	/** Maximum number of schedule queues to be supported
+	 * Application configures the maximum number of schedule queues to be
+	 * supported by the implementation.
+	 * @see odp_queue_capability_t
+	 */
+	uint32_t queue_count;
+
+	/** Maximum number of events required to be stored simultaneously in
+	 * schedule queue. This number must not exceed 'max_queue_size'
+	 * supported by the implementation.
+	 * A value of 0 configures default queue size supported by the
+	 * implementation.
+	 */
+	uint32_t queue_size;
+} odp_schedule_config_t;
+
+/**
+ * Scheduler capabilities
+ */
+typedef struct odp_schedule_capability_t {
+	/** Maximum supported flows per queue
+	 * Specifies the maximum number of flows per queue supported by the
+	 * implementation.
+	 * A value of 0 indicates flow aware mode is not supported.
+	 */
+	uint32_t max_flows;
+
+	/** Maximum supported schedule queues
+	 * Specifies the maximum number of schedule queues supported by the
+	 * implementation.
+	 */
+	uint32_t max_sched_queues;
+
+	/** Maximum number of events a schedule queue can store simultaneoulsy.
+	 * A value of 0 indicates the implementation does not restrict the
+	 * queue size
+	 */
+	uint32_t max_queue_size;
+
+	/** Maximum schedule group
+	 * Specifies the maximum number of schedule group supported by the
+	 * implementation.
+	 */
+	uint32_t max_sched_group;
+} odp_schedule_capability_t;
+
+/**
+ * Start scheduler operation
+ *
+ * Activate scheduler module to schedule events across different schedule
+ * queues. The scheduler module should be started before creating any odp
+ * queues. This should happen before an application itself or any other ODP
+ * module creates any schedued queue.
+ *
+ * The initialization sequeunce should be,
+ * odp_schedule_capability()
+ * odp_schedule_config_init()
+ * odp_schedule_config()
+ * odp_schedule_start()
+ * odp_schedule()
+ *
+ * @retval 0 on success
+ * @retval <0 on failure
+ */
+int odp_schedule_start(void);
```

Comment:
I'm still wondering why this API is needed. Aside from backwards compatibility issues (Do all existing apps need to change to add this call if they want to continue using the ODP scheduler? That really seems a negative.) the only reason to have this API is if the idea is the app will launch multiple threads that will immediately call `odp_schedule()` before configuration has completed, at which point these calls would be expected to stall until the (main?) thread calls `odp_schedule_start()`. Such stalling could be complicated to implement safely. If we're not going to allow this behavior, then it's just as easy to say that `odp_schedule_config()` should be called prior to any `odp_schedule()` call, and that's sufficient.

Reply:
When application can change config, some parts of scheduler global init needs to move later, but before the first schedule() call. So, a new step is needed on every application either it is schedule_config(NULL) or schedule_start(). I think it's more intuitive to keep config call optional and require only start() if application maintains the default config. It also enables later adding stop and re-config features if needed.

Reply:
How is it that the `odp_schedule_config()` call itself doesn't serve this purpose? If it's called, then a non-default set of configuration options is applied. If it isn't then the defaults remain.

What happens if `odp_schedule()` is called before `odp_schedule_config()` or `odp_schedule_start()`? Either we say that's not allowed, in which case there's no need for `odp_schedule_start()` as a sync point, or else the implication is that those calls(s) will stall until some other thread makes them, which is going to impose it's own implementation difficulties.

As far as reconfiguration goes, we already have `odp_schedule_pause()` and `odp_schedule_resume()` and we could consider `odp_schedule_stop()` and `odp_schedule_start()` to be global analogs of those thread-local calls. But note that `odp_schedule_resume()` is only needed to restart scheduling after `odp_schedule_pause()` is called. We do not require threads to call `odp_schedule_resume()` when they first start up. So we could have an `odp_schedule_start()` call that would only be needed after `odp_schedule_stop()`.

Reply:
@Bill-Fischofer-Linaro 
Without `odp_schedule_start() we will need:
- to check during each `odp_schedule()` call that scheduler was not configured/started and start it with default parameters.
- to disallow multiple `odp_schedule_config()` calls which can be used to negotiate number of flows.